File: contests/291_DIV2/e.py
# -*- coding: utf-8 -*-

import numpy as np
import copy

maxn = 100
MOD = 10**9 + 7
d = { i+1: 0 for i in xrange(maxn)}
n, x = map( int, raw_input().split() )

for i in raw_input().split():
  d[int(i)] += 1

A = [ [0] * (maxn+1) for _ in xrange(maxn+1) ]
X = [0] * maxn + [1]
X = [X]

for i in xrange(maxn-1):
  A[i+1][i] = 1
for i in xrange(maxn):
  A[i][maxn-1] = d[maxn-i]
A[maxn][maxn-1] = 1
A[maxn][maxn] = 1

A, ret, base = np.mat(A, dtype=np.uint64), np.mat(A, dtype=np.uint64), np.mat(A, dtype=np.uint64)

#ret = copy.deepcopy(A)
#base = copy.deepcopy(A)

def mul(a, b):
  ret = [ [0]*len(b[0]) for _ in xrange(len(a)) ]
  for i in xrange(len(a)):
    for j in xrange(len(b[0])):
      for k in xrange(len(b)):
        ret[i][j] += a[i][k] * b[k][j]
      ret[i][j] %= MOD
  return ret

while x:
  if x&1:
    #ret  = mul(ret, base)
    ret *= base  # TODO wil overflow when the number is huge.
    ret %= MOD
  #base = mul(base, base)
  base *= base
  base %= MOD
  x >>= 1

#ans = mul(X, ret)
ans = np.asarray(X*ret, dtype=np.uint64)
print int( ans[0][-2]%MOD )

File: contests/291_DIV2/e.cpp
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <iostream>

using namespace std;
typedef long long int64;

int n, x;
const int maxn = 100;
const int64 MOD = (int64)1e9 + 7;
int64 d[maxn+1];
int64 A[maxn+1][maxn+1];
int64 ret[maxn+1][maxn+1];
int64 base[maxn+1][maxn+1];
int64 tmp[maxn+1][maxn+1];
int64 X[1][maxn+1];


void mul(int64 a[][maxn+1], int64 b[][maxn+1]) {
  for (int i = 0; i < maxn+1; ++ i) {
    for (int j = 0; j < maxn+1; ++ j) {
      tmp[i][j] = 0;
      for (int k = 0; k < maxn+1; ++ k) {
        tmp[i][j] += a[i][k] * b[k][j];
        tmp[i][j] %= MOD;
      }
    }
  }
  memcpy(a, tmp, sizeof(tmp));
}

void solve() {
  cin >> n >> x;
  int64 t;
  memset(d, 0, sizeof(d));
  for (int i = 0; i < n; ++ i) {
    cin >> t;
    ++ d[t];
  }

  memset(A, 0, sizeof(A));
  memset(X[0], 0, sizeof(X[0]));
  X[0][maxn] = 1;

  for (int i = 0; i < maxn-1; ++ i) {
    A[i+1][i] = 1;
  }

  for (int i = 0; i < maxn; ++ i) {
    A[i][maxn-1] = d[maxn-i];
  }

  A[maxn][maxn-1] = 1;
  A[maxn][maxn] = 1;

  memcpy(ret, A, sizeof(A));
  memcpy(base, A, sizeof(A));

  while (x) {
    if (x&1) {
      mul(ret, base);
    }
    mul(base, base);
    x >>= 1;
  }

  int64 ans = 0;
  for (int i = 0; i < maxn+1; ++ i) {
    ans += X[0][i]* ret[i][maxn-1];
    ans %= MOD;
  }
  cout << ans << endl;
}

int main() {
  solve();
  return 0;
}
